pandas/core/strings.py

```diff
@@ -1815,36 +1817,34 @@ def __init__(self, data):
     def _validate(data):
         from pandas.core.index import Index
 
-        if (isinstance(data, ABCSeries) and
-                not ((is_categorical_dtype(data.dtype) and
-                      is_object_dtype(data.values.categories)) or
-                     (is_object_dtype(data.dtype)))):
-            # it's neither a string series not a categorical series with
-            # strings inside the categories.
-            # this really should exclude all series with any non-string values
-            # (instead of test for object dtype), but that isn't practical for
-            # performance reasons until we have a str dtype (GH 9343)
+        if isinstance(data, ABCMultiIndex):
+                raise AttributeError('Can only use .str accessor with Index, '
+                                     'not MultiIndex')
+
+        # see src/inference.pyx which can contain string values
+        allowed_types = ['string', 'unicode', 'empty',
+                         'mixed', 'mixed-integer']
+        if isinstance(data, ABCSeries):
+            allowed_types = allowed_types + ['bytes']
+
+        values = data if isinstance(data, Index) else data.values
+        if is_categorical_dtype(data.dtype):
+            inf_type = lib.infer_dtype(values.categories)
+        else:
+            inf_type = lib.infer_dtype(values)
+
+        all_na_obj = is_object_dtype(values.dtype) and data.isna().all()
```

Comment:
this is by defintion what 'empty' means, you only need to check if its object type in that case

Reply:
Unfortunately, that's not True. For all-NA Series/Index of object dtype, the inference still returns `'floating'`.
```
>>> import pandas as pd
>>> import numpy as np
>>> import pandas._libs.lib as lib
>>> lib.infer_dtype(pd.Index([np.nan, np.nan]))
'floating'
>>> lib.infer_dtype(pd.Series([np.nan, np.nan]))
'floating'
>>> lib.infer_dtype(pd.Series([]))
'floating'
>>> lib.infer_dtype(pd.Index([]))
'empty'
>>> lib.infer_dtype(pd.Index([np.nan, np.nan], dtype=object))
'floating'
>>> lib.infer_dtype(pd.Series([np.nan, np.nan], dtype=object))
'floating'
>>> lib.infer_dtype(pd.Series([], dtype=object))
'empty'
>>> lib.infer_dtype(pd.Index([], dtype=object))
'empty'
```

Reply:
ok, still separating these out to separate if/elif clauses makes this easier to grok.